activities/loadtest/loadtest.go

```diff
@@ -0,0 +1,367 @@
+		logger.Fatal("failed to recover faucet wallet key", zap.Error(err))
+	}
+	time.Sleep(1 * time.Second)
+
+	chainConfig := chain.GetConfig()
+	command := []string{
+		chain.GetConfig().BinaryName,
+		"tx", "bank", "multi-send",
+		faucetWallet.FormattedAddress(),
+	}
+
+	command = append(command, addresses...)
+	command = append(command, "1000000000stake",
+		"--chain-id", chainConfig.ChainId,
+		"--keyring-backend", "test",
+		"--fees", "3000stake",
+		"--gas", "auto",
+		"--yes",
+		"--home", chainConfig.HomeDir,
+	)
+
+	_, stderr, exitCode, err := node.RunCommand(ctx, command)
+	if err != nil || exitCode != 0 {
+		logger.Warn("failed to fund wallet", zap.Error(err), zap.String("stderr", stderr))
+	}
+	time.Sleep(5 * time.Second)
+
+	var msgs []Message
+	for _, msg := range loadTestConfig.Msgs {
+		msgs = append(msgs, Message{
+			Type:   msg.Type,
+			Weight: msg.Weight,
+		})
+	}
+
+	config := LoadTestConfig{
+		ChainID:             chainID,
+		BlockGasLimitTarget: loadTestConfig.BlockGasLimitTarget,
+		NumOfBlocks:         loadTestConfig.NumOfBlocks,
+		NodesAddresses:      nodes,
+		Mnemonics:           mnemonics,
+		GasDenom:            chain.GetConfig().Denom,
+		Bech32Prefix:        chain.GetConfig().Bech32Prefix,
+		Msgs:                msgs,
+	}
+	logger.Info("Load test config constructed", zap.Any("config", config))
+
+	return yaml.Marshal(&config)
+}
+
+func (a *Activity) RunLoadTest(ctx context.Context, chainState []byte,
+	loadTestConfig *LoadTestConfig, runnerType string, providerState []byte) (PackagedState, error) {
+	logger, _ := zap.NewDevelopment()
+
+	var p provider.ProviderI
+	var err error
+	if runnerType == string(testnettypes.Docker) {
+		p, err = docker.RestoreProvider(
+			ctx,
+			logger,
+			providerState,
+		)
+	} else {
+		p, err = digitalocean.RestoreProvider(
+			ctx,
+			providerState,
+			a.DOToken,
+			a.TailscaleSettings,
+			digitalocean.WithLogger(logger),
+		)
+	}
+
+	if err != nil {
+		return PackagedState{}, err
+	}
+
+	chain, err := chain.RestoreChain(ctx, logger, p, chainState, node.RestoreNode, testnet.CosmosWalletConfig)
+	if err != nil {
+		return PackagedState{}, err
+	}
+
+	configBytes, err := generateLoadTestConfig(ctx, logger, chain, chain.GetConfig().ChainId, loadTestConfig)
+	if err != nil {
+		return PackagedState{}, err
+	}
+
+	task, err := p.CreateTask(ctx, provider.TaskDefinition{
+		Name:          "catalyst",
+		ContainerName: "catalyst",
+		Image: provider.ImageDefinition{
+			Image: "ghcr.io/skip-mev/catalyst:latest",
+			UID:   "100",
+			GID:   "100",
+		},
+		ProviderSpecificConfig: map[string]string{
+			"region":   "ams3",
+			"image_id": "177032231",
+			"size":     "s-4vcpu-8gb",
+		},
+		Command: []string{"/tmp/catalyst/loadtest.yml"},
+		DataDir: "/tmp/catalyst",
+		Environment: map[string]string{
+			"DEV_LOGGING": "true",
+		},
+	})
+
+	if err != nil {
+		return PackagedState{}, err
+	}
+
+	if err := task.WriteFile(ctx, "loadtest.yml", configBytes); err != nil {
```

Comment:
we need to make sure that when we run into an error, we still serialize the provider, otherwise the state does not reflect the real world